README
NAME
    Mousse - A Light and Tasty Moose for CPAN Authors

SYNOPSIS
    In your "Makefile.PL":

        use inc::Module::Install;
        name 'Chocolate';
        use_mousse 'Chocolate::Mousse';

    then in "lib/Chocolate/Jesus.pm":

        package Chocolate::Jesus;
        use Chocolate::Mousse;

        has 'soul', is => 'ro';

    and:

        package Chocolate::Jeebus;
        use Chocolate::Mousse;

        extends 'Chocolate::Jesus';

        has 'whiskers', is => 'rw';

DESCRIPTION
    Mousse is the easiest way to get Moose support for your OO CPAN module.

    Mousse adds NO prerequisite dependencies to your CPAN module.

    Mousse has the same performance qualities as the zippy Mouse::Tiny.

    Mousse is enabled by adding one line to your Makefile.PL.

WHENCE MOUSSE?
    If you are a CPAN module author, you need to weigh the benefit of using
    Moose against the startup performance hit and sizable module
    prerequisite dependency. Should your simple module even use OO? Should
    you force a Moose dependency on every other module or project that uses
    your module? Should roll your own OO code and pass up the robustness of
    Moose? These questions grate on the concious of many CPAN authors.

    No longer!

    You've probably heard of Mouse. Mouse is a way to do all the core Moose
    functionality but without the performance costs. Mouse::Tiny puts all
    the Mouse modules into one module, but you still have a dependency on
    Mouse::Tiny. One solution is to copy Mouse::Tiny as your own module's OO
    base. But then you have to repeat that every time a new Mouse version is
    released.

    Another solution is to use Mousse! Mousse will remake your OO base
    module from the current version of Mousse have installed, every time you
    run "perl Makefile.PL".

MOUSSE FEATURES AND ADVANTAGES
  Mousse IS Mouse
    To learn about the features of Mousse (i.e. your generated Mousse-based
    OO module), simply read Mouse.

    Mousse is an exact copy of Mouse::*. It uses the same build process as
    Mouse::Tiny does, except that it names the resulting module whatever you
    want to call it. For example, if you use this line in your Makefile.PL:

        use_mousse 'Foo::Widget::OO';

    then Mousse::Maker will *whip up* a "lib/Foo/Widget/OO.pm" file for you
    from the Mousse.pm installed on your system.

    Alternatively, if you use *this* line in your Makefile.PL:

        use_mousse_dev 'Foo::Widget::OO';

    then Mousse::Maker will create a "lib/Foo/Widget/OO.pm" file for you
    from the *Mouse* parts installed on your system.

  No Dependency Hell Issues
    When you distribute a module that uses Mousse for OO, you can be assured
    that you have coded and tested against the exact copy of Mousse that
    your users will be running with. When a Mousse related bug is found, you
    can simply ship a new copy of your code (with the repaired copy of
    Mousse) and be certain that nobody will be trampled by a wild Moose.

  Using Mousse without Module::Install
    If you don't use Module::Install for your Makefile.PL, or if you want to
    only update Mousse by hand, you can use this simple command:

        perl -MMousse::Maker -e make_mousse 'Chocolate::Mousse' > lib/Chocolate/Mousse.pm

  You Can Use Mousse as a Prerequisite
    If you don't care about adding a dependency and you don't want to ship
    your own copy of Mousse, you can use it just like Mouse::Tiny.

EXAMPLES
    Scalar::Random::PP is an example of a CPAN module that uses Mousse.

    If you have a module to add here, let me know.

RESOURCES
    CPAN: <http://search.cpan.org/dist/Mousse/>

    GitHub: <http://github.com/ingydotnet/mousse-pm>

    IRC: irc.perl.org#moose-dev

CREDIT
    Thanks to Stevan Little for the movement that is Moose.

    The guts of this module is pure, un(well slightly)adulterated Mouse.
    藤吾郎++

    Also, many thanks to Luke Closs' cat, Jeebus!

AUTHOR
    Ingy döt Net <ingy@cpan.org>

COPYRIGHT
    Copyright (c) 2010. Ingy döt Net.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    See http://www.perl.com/perl/misc/Artistic.html



Changes
---
version: 0.14
date:    Wed Jun  1 15:04:30 EST 2011
changes:
- Build on Mouse-0.93
---
version: 0.13
date:    Sat May  7 21:55:48 EST 2011
changes:
- Build on Mouse-0.92
---
version: 0.12
date:    Sat Nov 13 19:45:13 PST 2010
changes:
- Build on Mouse-0.87
- Fix bugs in using without Module::Install (with tests!)
- Fixed bugs in Mousse::Maker
- Documented Mousse::Maker
- Removed overzealous Makefile.PL mousser
- Renamed foo.t to meaningful name -> module-install.t

---
version: 0.11
date:    Tue Oct  5 23:05:23 PDT 2010
changes:
- Build on Mouse-0.80
---
version: 0.10
date:    Sat Sep 18 07:13:52 PDT 2010
changes:
- Update docs
---
version: 0.07
date:    Fri Sep 17 15:35:53 PDT 2010
changes:
- Build on new Mouse-0.70
- Add manifest_skip and readme_from to Makefile.PL
- Updated docs
---
version: 0.06
date:    Thu Sep  9 17:57:44 PDT 2010
changes:
- use Mouse v0.65
- Change admin to is_admin in M::I::Mousse
- Fix doc
---
version: 0.05
date:    Tue Aug 31 19:14:54 PDT 2010
changes:
- t/foo.t failed without Module::Install
---
version: 0.04
date:    Tue Aug 31 08:50:47 PDT 2010
changes:
- Fun doc change per ilmari++
---
version: 0.03
date:    Sun Aug 29 18:41:20 PDT 2010
changes:
- Fixed another bootstrapping issue in Makefile.PL.
  Made dist with 'use_mousse_dev'.
---
version: 0.02
date:    Sun Aug 29 18:41:20 PDT 2010
changes:
- Fixed bootstrapping issue in Makefile.PL. Makefile.PL failed to find
  Mousse::Maker.
---
version: 0.01
date:    Thu Aug 26 06:35:25 PDT 2010
changes:
- Initial version.


t/mf.pl
use inc::Module::Install;

name 'Foo';

all_from 'lib/Foo.pm';

use_mousse 'Foo::Mousse';

# WriteAll;


t/lib/Foo.pm
package Foo;
use Foo::Mousse;
use 5.008003;

has 'this', is => 'rw';

1;

=head1 NAME

Foo - Testing the Mousse

=head1 AUTHOR

Ingy döt Net <ingy@cpan.org>

=head1 COPYRIGHT

Copyright (c) 2010. Ingy döt Net.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

See http://www.perl.com/perl/misc/Artistic.html

=cut


lib/Mousse/Maker.pm
package Mousse::Maker;
use strict;
use warnings;

use 5.008003;

use Exporter;
our @ISA = qw(Exporter);
our @EXPORT = qw(make_mousse);

our $VERSION = '0.14';

sub make_mousse {
    my $module = shift(@ARGV)
        or die "make_mousse requires a module name argument";
    die "'$module' doesn't look like a module name"
        unless $module =~ /^\w(\w|::)+$/;
    make_from_mousse($module, @ARGV);
}

sub make_from_mousse {
    my $Mousse = shift;
    my $MousseFile = shift || '';
    require Mousse;
    my $file = $INC{'Mousse.pm'};
    my $contents = slurp($file);

    my $Mousse2 = $Mousse;
    $Mousse2 =~ s/::/\//g;

    $contents =~ s/Mousse(.)/(($1 eq '\/' or $1 eq '.') ? $Mousse2 : $Mousse) .$1/ge;
    my $handle;
    if ($MousseFile) {
        my $MousseDir = $MousseFile;
        $MousseDir =~ s/(.*[\\\/]).*/$1/ and File::Path::mkpath($MousseDir);
        open $handle, ">$MousseFile"
            or die "Can't open $MousseFile for output";
    }
    else {
        $handle = \*STDOUT;
    }

    print { $handle } $contents;
}

# This is a modified version of tool/generate-mouse-tiny.pl
# from git://git.moose.perl.org/Mouse.git 
sub make_from_mouse {
    my $Mousse = shift;
    my $MousseFile = shift || '';
    require Mouse;
    my $lib = $INC{'Mouse.pm'};
    $lib =~ s/[\\\/]Mouse\.pm$//;

    use File::Find;
    use Fatal qw(open close);
    use File::Path;

    sub uniq{
        my %seen;
        return grep{ !$seen{$_}++ } @_;
    }

    require "$lib/Mouse/Spec.pm";

    my @files;

    find({
        wanted => sub {
            push @files, $_
                if -f $_
                &&  /\.pm$/
                &&  /\bMouse\b/
                & !/Squirrel/
                && !/Tiny/
                && !/Test/         # only for testing
                && !/Spec/         # has no functionality
                && !/TypeRegistry/ # deprecated
                && !/\bouse/       # ouse.pm
        },
        no_chdir => 1,
    }, $lib);

    my $mousse = '';

    for my $file (uniq
            "$lib/Mouse/PurePerl.pm",
            "$lib/Mouse/Exporter.pm",
            "$lib/Mouse/Util.pm",
            "$lib/Mouse/Meta/TypeConstraint.pm",
            "$lib/Mouse/Util/TypeConstraints.pm",
                sort @files) {

        my $module = $file;
        $module =~ s/^\Q$lib\E\/(.*)\.pm/$1/;
        $module =~ s/\//::/g;
        my $contents = slurp($file);

        fixups($module, \$contents);

        $contents =~ s/__END__\b.*//s;          # remove documentation
        $contents =~ s/1;\n*$//;                # remove success indicator

        $contents =~ s/Mouse/$Mousse/g;

        $contents =~ s{^( (?:[ ]{4})+ )}{ "\t" x (length($1) / 4) }xmsge; # spaces to tabs

        $mousse .= "# Contents of $module\n";
        $mousse .= $contents;
    }

    my $handle;
    if ($MousseFile) {
        open $handle, ">$MousseFile";
    }
    else {
        $handle = \*STDOUT;
    }

    print { $handle } << "EOF";
# This file was generated by Mousse::Maker $VERSION from Mouse $Mouse::Spec::VERSION.
#
# ANY CHANGES MADE HERE WILL BE LOST!
use strict;
use warnings;

# tell Perl we already have all of the Mousse files loaded:
EOF

    print { $handle } "BEGIN {\n";
    for my $file (@files) {
        $file =~ s/Mouse/$Mousse/g;
        (my $inc = $file) =~ s{^\Q$lib\E/}{};
        printf { $handle } "\t%-45s = __FILE__;\n", "\$INC{'$inc'}";
    }

    print { $handle } <<"END_OF_MOUSSE";
}

# and now their contents

$mousse;

package $Mousse;

our \$VERSION = '$Mousse::Maker::VERSION';

$Mousse\::Exporter->setup_import_methods(also => '$Mousse\::TOP');

1;
END_OF_MOUSSE

    close $handle;
}

sub fixups {
    my ($module, $contents) = @_;
    if ($module eq 'Mouse') {
        $$contents =~ s/Mouse/Mouse::TOP/m or die;
    }
    elsif (
        $module eq 'Mouse::Util' or
        $module eq 'Mouse::Meta::Module' or
        $module eq 'Mouse::Meta::Class'
    ) {
        $$contents =~ s/\n\n/\nno warnings 'once';\n\n/m or die;
    }
}

sub slurp {
    open my $in, '<', $_[0];
    local $/;
    return scalar <$in>;
}

1;

=encoding utf-8

=head1 NAME

Mousse::Maker - Make Your Own Mousse Module

=head1 SYNOPSIS

In your C<Makefile.PL>:

    use inc::Module::Install;
    name 'Chocolate';
    use_mousse 'Chocolate::Mousse';

or (if you don't use Module::Install):

    perl -MMousse::Maker -e make_mousse 'Chocolate::Mousse' > lib/Chocolate/Mousse.pm

=head1 DESCRIPTION

Mousse::Maker is the module that Module::Install::Mousse uses to generate your own Mousse OO base class module. You can also use it by hand in a Perl one liner (shown above) if your CPAN module doesn't use Module::Install for its Makefile.PL.

=head1 AUTHOR

Ingy döt Net <ingy@cpan.org>

=head1 COPYRIGHT

Copyright (c) 2010, 2011. Ingy döt Net.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

See http://www.perl.com/perl/misc/Artistic.html

=cut
